File: apps/mobile/android/app/src/main/java/com/cozypad/app/SshCredentialVault.kt
package com.cozypad.app

import java.util.concurrent.ConcurrentHashMap

internal data class SshTarget(
    val host: String,
    val port: Int,
    val username: String,
)

internal sealed interface SshCredential {
    val target: SshTarget
    val authMethod: String

    data class Password(
        override val target: SshTarget,
        val password: String,
    ) : SshCredential {
        override val authMethod = "password"
    }

    data class PrivateKey(
        override val target: SshTarget,
        val privateKey: String,
        val passphrase: String?,
    ) : SshCredential {
        override val authMethod = "privateKey"
    }
}

internal interface SshCredentialPersistence {
    fun get(profileId: String): SshCredential?
    fun put(profileId: String, credential: SshCredential)
    fun remove(profileId: String)
}

internal data class SshCredentialStatus(
    val hasCredential: Boolean,
    val persisted: Boolean,
)

/**
 * Keeps SSH secrets outside the WebView. Persisted credentials are encrypted by
 * [AndroidSshCredentialPersistence]; non-persisted credentials live only for
 * the current native process so reconnect remains seamless.
 *
 * Credentials are bound to host, port, username, and authentication method.
 * A compromised renderer therefore cannot redirect a saved password to a
 * different SSH server merely by changing profile metadata.
 */
internal class SshCredentialVault(
    private val persistence: SshCredentialPersistence,
) {
    private val transient = ConcurrentHashMap<String, SshCredential>()

    @Synchronized
    fun configure(
        profileId: String,
        target: SshTarget,
        authMethod: String,
        remember: Boolean,
        supplied: SshCredential?,
    ): Boolean {
        requireValid(profileId, target, authMethod)
        if (supplied != null) {
            require(supplied.target == target && supplied.authMethod == authMethod) {
                "credential does not match its SSH profile"
            }
            if (remember) {
                persistence.put(profileId, supplied)
                transient.remove(profileId)
            } else {
                persistence.remove(profileId)
                transient[profileId] = supplied
            }
            return true
        }

        val transientCredential = transient[profileId]
        if (transientCredential != null) {
            if (
                transientCredential.target != target ||
                transientCredential.authMethod != authMethod
            ) {
                remove(profileId)
                return false
            }
            if (remember) {
                persistence.put(profileId, transientCredential)
                transient.remove(profileId)
            }
            return true
        }

        val persistedCredential = persistence.get(profileId)
        if (
            persistedCredential == null ||
            persistedCredential.target != target ||
            persistedCredential.authMethod != authMethod
        ) {
            remove(profileId)
            return false
        }

        if (!remember) {
            remove(profileId)
            return false
        }
        return true
    }

    @Synchronized
    fun get(
        profileId: String,
        target: SshTarget,
        authMethod: String,
    ): SshCredential? {
        requireValid(profileId, target, authMethod)
        val credential = transient[profileId] ?: persistence.get(profileId) ?: return null
        return credential.takeIf {
            it.target == target && it.authMethod == authMethod
        }
    }

    @Synchronized
    fun status(
        profileId: String,
        target: SshTarget,
        authMethod: String,
    ): SshCredentialStatus {
        requireValid(profileId, target, authMethod)
        transient[profileId]?.let {
            if (it.target == target && it.authMethod == authMethod) {
                return SshCredentialStatus(hasCredential = true, persisted = false)
            }
        }
        persistence.get(profileId)?.let {
            if (it.target == target && it.authMethod == authMethod) {
                return SshCredentialStatus(hasCredential = true, persisted = true)
            }
        }
        return SshCredentialStatus(hasCredential = false, persisted = false)
    }

    @Synchronized
    fun remove(profileId: String) {
        transient.remove(profileId)
        persistence.remove(profileId)
    }

    private fun requireValid(profileId: String, target: SshTarget, authMethod: String) {
        require(profileId.isNotBlank() && profileId.length <= 256) { "invalid profileId" }
        require(target.host.isNotBlank() && target.host.length <= 255) { "invalid SSH host" }
        require(target.port in 1..65535) { "invalid SSH port" }
        require(target.username.isNotBlank() && target.username.length <= 256) {
            "invalid SSH username"
        }
        require(authMethod == "password" || authMethod == "privateKey") {
            "unsupported authMethod: $authMethod"
        }
    }
}

File: packages/capacitor-ios-native/Sources/CozypadCapacitorIosNative/SshCredentialVault.swift
import Foundation

struct SshTarget: Codable, Equatable, Sendable {
    let host: String
    let port: Int
    let username: String
}

struct SshCredential: Codable, Equatable, Sendable {
    let target: SshTarget
    let authMethod: String
    let password: String?
    let privateKey: String?
    let passphrase: String?
}

struct SshCredentialStatus {
    let hasCredential: Bool
    let persisted: Bool
}

final class SshCredentialVault: @unchecked Sendable {
    private let keychain: KeychainStore
    private let lock = NSLock()
    private var transient: [String: SshCredential] = [:]

    init(keychain: KeychainStore) {
        self.keychain = keychain
    }

    func configure(
        profileId: String,
        target: SshTarget,
        authMethod: String,
        remember: Bool,
        supplied: SshCredential?
    ) throws -> SshCredentialStatus {
        try validate(profileId: profileId, target: target, authMethod: authMethod)
        return try lock.withLock {
            if let supplied {
                guard supplied.target == target, supplied.authMethod == authMethod else {
                    throw VaultError.invalidCredential
                }
                if remember {
                    try persist(supplied, profileId: profileId)
                    transient.removeValue(forKey: profileId)
                } else {
                    try keychain.remove(storageKey(profileId))
                    transient[profileId] = supplied
                }
                return SshCredentialStatus(hasCredential: true, persisted: remember)
            }

            if let existing = transient[profileId] {
                guard existing.target == target, existing.authMethod == authMethod else {
                    try removeLocked(profileId)
                    return SshCredentialStatus(hasCredential: false, persisted: false)
                }
                if remember {
                    try persist(existing, profileId: profileId)
                    transient.removeValue(forKey: profileId)
                    return SshCredentialStatus(hasCredential: true, persisted: true)
                }
                return SshCredentialStatus(hasCredential: true, persisted: false)
            }

            guard let existing = try persisted(profileId) else {
                return SshCredentialStatus(hasCredential: false, persisted: false)
            }
            guard existing.target == target, existing.authMethod == authMethod, remember else {
                try removeLocked(profileId)
                return SshCredentialStatus(hasCredential: false, persisted: false)
            }
            return SshCredentialStatus(hasCredential: true, persisted: true)
        }
    }

    func credential(profileId: String, target: SshTarget, authMethod: String) throws -> SshCredential? {
        try validate(profileId: profileId, target: target, authMethod: authMethod)
        return try lock.withLock {
            let value = transient[profileId] ?? (try persisted(profileId))
            guard value?.target == target, value?.authMethod == authMethod else { return nil }
            return value
        }
    }

    func status(profileId: String, target: SshTarget, authMethod: String) throws -> SshCredentialStatus {
        try validate(profileId: profileId, target: target, authMethod: authMethod)
        return try lock.withLock {
            if let value = transient[profileId], value.target == target, value.authMethod == authMethod {
                return SshCredentialStatus(hasCredential: true, persisted: false)
            }
            if let value = try persisted(profileId), value.target == target, value.authMethod == authMethod {
                return SshCredentialStatus(hasCredential: true, persisted: true)
            }
            return SshCredentialStatus(hasCredential: false, persisted: false)
        }
    }

    func remove(_ profileId: String) throws {
        try lock.withLock { try removeLocked(profileId) }
    }

    private func validate(profileId: String, target: SshTarget, authMethod: String) throws {
        guard !profileId.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty,
              profileId.count <= 256 else { throw VaultError.invalidProfile }
        guard !target.host.isEmpty, target.host.count <= 255,
              (1...65_535).contains(target.port),
              !target.username.isEmpty, target.username.count <= 256 else {
            throw VaultError.invalidTarget
        }
        guard authMethod == "password" || authMethod == "privateKey" else {
            throw VaultError.unsupportedAuthentication
        }
    }

    private func persisted(_ profileId: String) throws -> SshCredential? {
        guard let data = try keychain.data(for: storageKey(profileId)) else { return nil }
        do {
            return try JSONDecoder().decode(SshCredential.self, from: data)
        } catch {
            try keychain.remove(storageKey(profileId))
            return nil
        }
    }

    private func persist(_ credential: SshCredential, profileId: String) throws {
        try keychain.set(JSONEncoder().encode(credential), for: storageKey(profileId))
    }

    private func removeLocked(_ profileId: String) throws {
        transient.removeValue(forKey: profileId)
        try keychain.remove(storageKey(profileId))
    }

    private func storageKey(_ profileId: String) -> String { "ssh-credential:\(profileId)" }
}

enum VaultError: Error, LocalizedError {
    case invalidProfile
    case invalidTarget
    case invalidCredential
    case unsupportedAuthentication

    var errorDescription: String? {
        switch self {
        case .invalidProfile: return "invalid profileId"
        case .invalidTarget: return "invalid SSH target"
        case .invalidCredential: return "credential does not match its SSH profile"
        case .unsupportedAuthentication: return "unsupported authentication method"
        }
    }
}
